node/src/components/linear_chain.rs

```diff
@@ -286,60 +335,135 @@ where
+            Event::GetBlockForFinalitySignaturesResult(fs, maybe_block) => {
+                if let Some(block) = &maybe_block {
+                    assert_eq!(
+                        block.hash(),
+                        &fs.block_hash,
+                        "block loaded from storage should have a matching block hash."
                     )
                 }
-                Effects::new()
+                // Check if validator is bonded in the era in which the block was created in.
+                effect_builder
+                    .is_bonded_validator(fs.era_id, fs.public_key)
+                    .map(|is_bonded| {
+                        if is_bonded {
+                            Ok((maybe_block, fs, is_bonded))
+                        } else {
+                            // If it's not bonded in that era, we will check if it's bonded in the
+                            // future era.
+                            Err((maybe_block, fs))
+                        }
+                    })
             }
-            Event::GetBlockForFinalitySignaturesResult(_block_hash, Some(block)) => {
-                let old_count = block.proofs().len();
-                let (block, mut effects) =
-                    self.add_pending_finality_signatures(*block, effect_builder);
-                if block.proofs().len() > old_count {
-                    let block = Box::new(block);
-                    effects.extend(effect_builder.put_block_to_storage(block).ignore());
+            .result(
+                |(maybe_block, fs, is_bonded)| Event::IsBonded(maybe_block, fs, is_bonded),
+                |(maybe_block, fs)| Event::IsBondedFutureEra(maybe_block, fs),
+            ),
+            Event::IsBondedFutureEra(maybe_block, fs) => {
+                // This should be safe to do. `self.latest_block` will be None only if ew haven't
+                // finalized *any* block yet. which will be just after the Genesis
+                // era start. Even if we receive a finality signature before finalizing anything,
+                // that will be for the *current* era and we won't be looking up whether validator
+                // is bonded in the Contract Runtime.
+                let state_root_hash = self
+                    .latest_block
+                    .as_ref()
+                    .map(|lb| *lb.header().state_root_hash())
+                    .unwrap();
+                let next_era_id = self
+                    .latest_block
+                    .as_ref()
+                    .map(|lb| lb.header().era_id().successor())
+                    .unwrap();
+                // TODO: Use protocol version that is valid for the block's height.
+                let protocol_version = ProtocolVersion::new(SemVer::V1_0_0);
+                effect_builder
+                    .is_bonded_in_future_era(
+                        state_root_hash,
+                        next_era_id,
```

Comment:
Why not `fs.era_id`? It could be for a block further ahead — wouldn't we be looking up the wrong validator set here?

Reply:
`fs.era_id` and block's `era_id` should much though?

Reply:
Why? `era_id` is our own local latest block's era ID, and the finality signature could apply to a completely different block, couldn't it?

Reply:
Ah right :facepalm: but since we already checked that it wasn't created in the current era, we want to test if it was created in the next era.

Reply:
It could be two eras ahead.
If we want to just reject it in that case, we should still check `fs.era_id` to find out.

Reply:
> It could be two eras ahead.

Yes, but I only check for 1 era ahead.

Reply:
So won't we get a false positive if the finality signature is for a block two eras ahead?
Say we're in era 5, the signature is for a block in era 7, and the signer actually isn't a validator in era 7, but was in era 6 — wouldn't we wrongly accept the signature then?

Reply:
Hm, you're right. I should use the `era_id` from the signature.

Reply:
Done in [`3bffb83` (#683)](https://github.com/CasperLabs/casper-node/pull/683/commits/3bffb831458819c2ed21e895cd3fcf945c06f1a4)